### PR TITLE
[SPARK-4681] Enable executor blacklisting by default.

### PR DESCRIPTION
**Pull request by kayousterhout**
This commit sets the default value of spark.scheduler.executorTaskBlacklistTime
to 5 seconds, so that if a task fails on a particular executor, it won't
be re-launched on the same executor until 5 seconds has elapsed, to allow
time for the task to (preferably) be launched on a different executor, where
it may be more likely to complete successfully.

One question here is whether we should now document this configuration option,
under advanced options?

This enabling is as-per the very old JIRA and the discussion on #8760, and should be
merged in 1.6.

@davies let me know whether you think this seems like an appropriate default?

cc @squito

### CODE REVIEW

**Comment by kayousterhout** [comment]
and cc @mridulm who originally wrote this feature

**Comment by davies** [comment]
@kayousterhout I'd like to increase the timeout to a little bit larger, for example, one minute.

Considering that we have N executors and M tasks (M > N), if there is one bad executor (bad host), all other executors will be busy with computing, other idle tasks will be scheduled to this bad executor and fail. If the timeout * 3 is smaller than the time used to finished all the task in (N-1) executor, the job will fail.

The disadvantage of having larger timeout is that you have only one executor and hit some temporary failure (could succeed after retry), it will slow down the job.

**Comment by kayousterhout** [comment]
@davies Hmmm I wonder if the best solution here is to change the blacklist behavior, so that as long as there are more available executors, we just permanently blacklist any executor where the task has already failed. I realize this is a more invasive change, but seems like it would lead to better default behavior.  Thoughts?

**Comment by davies** [comment]
@kayousterhout The number of available executors could change all the time, I'm thinking of a small change that permanently blacklist a executor for second failure, we never schedule a task to a executor, if the task had failed on the executor twice.

The patch could be small,:
```
-    failedExecutors.getOrElseUpdate(index, new HashMap[String, Long]()).
-      put(info.executorId, clock.getTimeMillis())
+    val failed = failedExecutors.getOrElseUpdate(index, new HashMap[String, Long]())
+    if (failed.contains(info.executorId)) {
+      // never try the executor again if the task failed on it twice
+      failed.put(info.executorId, Int.MaxValue)
+    } else {
+      failed.put(info.executorId, clock.getTimeMillis())
+    }
```

**Comment by kayousterhout** [comment]
@davies I like that approach, but we'd also need to add a check to make sure the job won't hang, right? Since we could end up in a situation where there is only 1 executor, so the job's remaining task(s) can't run on it, but the job won't have hit the 3-failure abort threshold.

**Comment by davies** [comment]
Good point, we may also want to fail fast in local cluster (having only one executor). Ideally, the tasks should be retried on other failure domains (other executors) first, then back to previous one. The difficult part is that the list of executors is dynamic. To simplify this, we can assume that the list of executors is static when schedule a task (or retry). So your suggestion is better (the timeout behavior depends on the number of available executors).

**Comment by mridulm** [comment]
@kayousterhout I would have preferred if this "feature" had been actually fixed properly by now - it was added as a temporary workaround (and so intentionally not documented so that we can remove it later) to an immediate problem for which we did not have a principled solution for at that time.

@davies The proposed change (to infinitely blacklist an executor) is not unsound IMO. Task failures are usually due to transient reasons - temporary disk or memory limitation, etc - infinitely blacklisting makes assumption about locality level timeouts, cost-benefit of retrying vs moving RDD partition to other node, etc.

On other hand, I do see user's complaining about needing to blacklist nodes as opposed to executors fairly commonly - which is something we have seen is an issue too (bad memory or disk for example causing most tasks on a node to fail).


It would be great if the current hack is removed entirely and we had a better more principled solution to this problem which handles executor, node and rack (required ?) blacklisting - rather than duct tape it further : but since I am not an active contributor off late, I might be oversimplifying the effort required :-) Something similar to what I did for task scheduling perhaps ?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 1]
```diff
@@ -58,14 +58,21 @@ private[spark] class TaskSetManager(
...
-  /*
-   * Sometimes if an executor is dead or in an otherwise invalid state, the driver
-   * does not realize right away leading to repeated task failures. If enabled,
-   * this temporarily prevents a task from re-launching on an executor where
-   * it just failed.
+  /**
+   * This timeout (specified in milliseconds) is used to prevent tasks from being immediatley
+   * re-launched on an executor where the task has already failed. A task will not be re-launched
+   * on an executor where it has already failed until this amount of time has elapsed since the
+   * failure.  One example of when this is useful is if an executor is in dead and the driver
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: extra "in".  (or maybe it was supposed to be "in fact")

I might change to talk about misconfigured executors, that appear to be alive, but can't run tasks.  At least thats the use case I've seen for this.  But no big deal really.

**Comment by kayousterhout** [comment]
After thinking about this (and some of the corner cases) more I would propose two steps here:

(1) For blacklisting executors per-task, I think @davies proposal makes most sense, which was to allow a task to run twice on a particular executor, and if it fails both times on that executor, blacklist the executor for the task permanently.  The benefit of that proposal is that it works without requiring the user to configure a blacklist timeout; determining a timeout is very workload dependent, and I think puts an unnecessary burden on the user.  I think this handles the transient failure issue relatively well, since the executor is only blacklisted for the particular task that failed twice, so there are plenty of future opportunities (with other tasks) for the executor to keep running things.  I do think it makes sense to have a per-task blacklist that's more aggressive than the more general executor/host blacklist mechanism.  @mridulm does this seem reasonable?

(2) We still need a solution for handling executors / hosts that are more permanently problematic, in which case the above approach of blacklisting only for a particular task is too inefficient.  Let's have that discussion on #8760, to separate the two issues.

**Comment by squito** [comment]
I think this change is fine as a "safe" change to make immediately.  I have been hesitant to change the default behavior, I figured there was some reason it was turned off so far.  But, fwiw, I always advice customers with large clusters to use this, and to crank that value up a lot more.  5s seems like a better default, though, without too aggressively changing the existing behavior.

I don't think this solves the larger problem of having an improved blacklisting mechanism, but I think that discussion goes elsewhere, assuming you want to stick with this simple change here.

**Comment by squito** [comment]
sorry race in our comments.  I agree that configuring timeouts is a real pain, and like the idea of some set number of times per executor.

it is a bigger change, though, and I'm wondering if it makes sense to just do the change proposed here if you want something in 1.6.

**Comment by kayousterhout** [comment]
I like the idea of fixing this for 1.6, but based on the discussion above, it seems like the appropriate timeout varies widely, so I'm worried if we set the timeout high enough to be effective, it may also lead to big performance issues for other people (and I'm also hesitant to "expose" this feature more if we're about to change it)

**Comment by squito** [comment]
makes sense.  I'm similarly hesistant to expose and then completely rewrite (I forgot to respond to that question of yours initially on documenting the timeout -- I'd like to expose whatever the eventual configuration options are, but also don't want to doc it know if we're about to change it.)

**Comment by mridulm** [comment]
@kayousterhout Permanently disabling task on an executor can be problematic for a few reasons :
a) When locality timeout's are aggressively high (user config, so we dont control it).
b) When it is very suboptimal to move task from the executor (usually (a)) - particularly relevant when data is stored off-VM, like in accelerator card, gpu, etc.
c) Any infinite timeout is risky - a very aggresively high default value which can be controlled by user would be my preference to permanently disabling it.
Having said that, as I mentioned before, would prefer if we just eliminated this - I was never expecting this to survive for this long.

(b) actually results in a lot of task failures due to transient issues (particularly at dev time) - but the cost of moving data out is extremely expensive.

**Comment by kayousterhout** [comment]
@mridulm these reasons make sense, but it's hard for me to imagine a case where a task has failed twice on a particular executor and the third time works (it seems like giving it a second chance should be good enough for transient issues). Are you saying that you have seen cases where the first two tries on an executor fail and the 3rd succeeds?

**Comment by mridulm** [comment]
@kayousterhout I have a job with 200k tasks where some tasks fail for as much as 22 times (no kidding - actual number) and then succeeds :-)

Specific examples (not sure if all of them are relevant) :

a) From an earlier work, GPU resource exhaustion would fail the task - but it would succeed if rescheduled.
b) Reducers failing due to fetch failures but succeed on reschedule (preferred locality set btw - so gets rescheduled on same node).
c) JNI related failures are fairly common in some of the libraries we use.
d) Direct buffer exhaustion which works on reschedule (race between tasks).

In some of these (a, b) the cost of moving the data to a different node is fairly expensive.

**Comment by squito** [comment]
:open_mouth: wow that is definitely not a situation I had given a lot of thought to :)
But would you agree this is a special case, behavior that a user would have to deal w/ configuration for, rather than the standard out of the box experience?

**Comment by mridulm** [comment]
Sure, I am fine with it not being default behavior - but there should be some way to override the default behavior. For example here, as long as there exists a way to override the number of failures (to say 128) I am fine with defaulting to infinite blacklist after 2.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 2]
```diff
@@ -58,14 +58,21 @@ private[spark] class TaskSetManager(
...
+   * this to 0.
+   *
+   * The motivation for the default value of 5 seconds is that it is longer than the default
+   * locality wait time (spark.locality.wait), so the task will be launched on an executor with
+   * worse locality (if one is available) before being re-launched on an executor where it failed
+   * (since running more slowly is preferable to failing).
    */
   private val EXECUTOR_TASK_BLACKLIST_TIMEOUT =
-    conf.getLong("spark.scheduler.executorTaskBlacklistTime", 0L)
+    conf.getLong("spark.scheduler.executorTaskBlacklistTime", 5000L)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: this should use `conf.getTimeAsSeconds` instead, though we should still accept the old integer value for backward compatibility

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: never mind, it's not even documented so we don't have to handle backward compatibility

**Comment by kayousterhout** [comment]
Sorry I let this lapse... closing as per the discussion above, that this blacklist behavior should really just be re-written, rather than exposed in it's current state.

**Pull request closed without merging**